### ISSUE TITLE
Auto prepend help

### ISSUE THREAD

**Issue by eurobank** [CIDRAM/CIDRAM#106]
I already use a prepend for my site (in htaccess), for the stand-alone Ninja Firewall. Adding a second prepend for CIDRAM doesn't seem to work, for some reason. Actually it seems that only the FIRST prepend (whatever it is) works. The Second prepend is ... ignored. 

So i'm using the loader of CIDRAM in one of my website early file, injecting the php code in the first line with no CR (like we did in ZBBLOCK). And works great.

But i want to do something better, still i'm not a programmer but an experienced webmaster.

I have created a php file and i put in there both loaders (Ninja & CIDRAM). Now not sure how to code this.

I created this (after an example i found), not sure if it is correctly coded.

```PHP
<?php
// Prepend CIDRAM for domain:
require ('/home/public_html/cidram/loader.php');

// Prepend the FIREWALL NINJA for domain
require ('/home/public_html/ninja/firewall.php');
``` 

While it seems to work, i'm puzzled by the missing of closing the php at the end. Is that right ? Also is "require" ok ? Or require once ?

Thanks

**Comment by Maikuolan** [comment]
> While it seems to work, i'm puzzled by the missing of closing the php at the end. Is that right ?

It doesn't really matter whether there's a closing tag `?>` at the end of the file, in most cases. In the past (i.e., the early days of PHP), a closing tag `?>` was often considered a good formality to adhere to, and a sign of good coding, even in cases where it wouldn't make any difference to the result of executing the code in question. Nowadays though, most people within the PHP ecosystem consider that the "best practice" in regards to PHP closing tag `?>` is (1) to include it if the PHP file in question mixes together content intended for the PHP interpreter (i.e., PHP code) with content not intended for the PHP interpreter (e.g., raw HTML data, template data, etc), and (2) to omit it if the PHP file contains only content intended for the PHP interpreter (e.g., if it contains only PHP code, doesn't directly produce any output itself, doesn't contain any template data, etc). This "best practice" also aligns closely with [PSR-2's stance regarding PHP closing tags](https://www.php-fig.org/psr/psr-2/#22-files):

> The closing `?>` tag MUST be omitted from files containing only PHP.

Whether you include it or omit it, either way, it isn't really "right" or "wrong". Best practice aside, there's an element of stylistic choice there. But, for all the PHP files in CIDRAM, I omit closing tags, primarily because I try to adhere to the PSR-2 coding style guide as closely as possible (and per their stance, because CIDRAM's PHP files are pure PHP files, intended wholly for the PHP interpreter, closing tags should be omitted).

> Also is "require" ok ? Or require once ?

Using `require` is perfectly okay. :-)

The only difference between `require` and `require_once` is that, whenever you use `require` on some particular file, it will be loaded into the instance, regardless of how many times you do this, whereas if you use `require_once` multiple times on some particular file, it will only be loaded once (at the first occurrence of the `require` or `require_once`). After a PHP file has been loaded (for a particular request/instance), subsequent calls to it using `require` will be honoured, but subsequent calls to it using `require_once` will not be honoured.

In short: If your main loader (the file you've written, for the purpose of loading both CIDRAM and NinjaFirewall together) is only executed once per request, it doesn't matter whether you use `require` or `require_once`. But, if there's a chance that your main loader could be executed multiple times per request (e.g., weird server configuration, accidentally calling it via multiple, linked entry points, etc), `require_once` could possibly be a safer choice, to prevent CIDRAM and NinjaFirewall from needlessly executing multiple times within the one request, slightly lightening the amount of work and pressure on your server/processor/RAM/etc. :-)

**Comment by eurobank** [comment]
Thanks for all the details. So i'm good i guess.
